### PR TITLE
chore(driver): arm deploy 

### PR DESCRIPTION
**Pull request by xzf0587**

### CODE REVIEW

File: packages/fx-core/src/component/driver/arm/error/invalidParameterUserError.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { UserError } from "@microsoft/teamsfx-api";
+import { getDefaultString, getLocalizedString } from "../../../../common/localizeUtils";
+
+const errorCode = "InvalidParameter";
+const messageKey = "driver.arm.error.invalidParameter";
```
Comment (blackchoey) [inline packages/fx-core/src/component/driver/arm/error/invalidParameterUserError.ts:8]: seems you didn't include package.nls.json

File: packages/fx-core/src/component/driver/arm/error/invalidParameterUserError.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+export class InvalidParameterUserError extends UserError {
```
Comment (blackchoey) [inline packages/fx-core/src/component/driver/arm/error/invalidParameterUserError.ts:10]: You may reuse similar error in AAD driver (and move it out of AAD folder). The wording already considers supporting different drivers.

File: packages/fx-core/src/component/driver/arm/util/util.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+function convertOutput(output: deploymentOutput, map: Map<string, string>, prefix?: string) {
+  const keys = Object.keys(output);
+  for (const key of keys) {
+    const value = output[key].value;
+    if (value instanceof Object) {
+      const newPrefix = buildKey(key, prefix);
+      convertOutput(value, map, newPrefix);
+    } else {
+      const mapKey = buildKey(key, prefix);
+      map.set(mapKey.toUpperCase(), value.toString());
```
Comment (blackchoey) [inline packages/fx-core/src/component/driver/arm/util/util.ts:41]: should we throw error / warning if the key already exists in map?

File: packages/fx-core/src/component/driver/arm/util/util.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+export function convertOutputs(outputs: deploymentOutput[]): Map<string, string> {
+  const res = new Map<string, string>();
+  for (const output of outputs) {
+    convertOutput(output, res);
+  }
+  return res;
+}
+
+function convertOutput(output: deploymentOutput, map: Map<string, string>, prefix?: string) {
```
Comment (blackchoey) [inline packages/fx-core/src/component/driver/arm/util/util.ts:32]: minor: may add a comment to describe the naming conversion rule for the output keys

File: packages/fx-core/src/component/driver/arm/error/invalidParameterUserError.ts (hunk 2 above)
Comment (xzf0587) [reply packages/fx-core/src/component/driver/arm/error/invalidParameterUserError.ts]: It is fixed. Thanks

File: packages/fx-core/src/component/driver/arm/util/util.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  const keys = Object.keys(output);
+  for (const key of keys) {
+    const value = output[key].value;
+    if (value instanceof Object) {
+      const newPrefix = buildKey(key, prefix);
+      convertOutput(value, map, newPrefix);
+    } else {
+      const mapKey = buildKey(key, prefix);
+      if (map.get(mapKey)) {
+        throw new Error(`There is duplicated key ${mapKey} in arm deployment output`);
```
Comment (jayzhang) [inline packages/fx-core/src/component/driver/arm/util/util.ts:62]: Better to extract as localized string

File: packages/fx-core/src/component/driver/arm/validator.ts [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+  } else {
+    // check the bicep cli version
+  }
+  return true;
+}
+
+function validateTemplates(templates: templateArgs[]): string[] {
+  const res: string[] = [];
+  if (templates.length === 0) {
+    res.push(`templates is invalid for it is empty`);
```
Comment (jayzhang) [inline packages/fx-core/src/component/driver/arm/validator.ts:50]: Better to extract as localized string

File: packages/fx-core/src/component/driver/arm/validator.ts [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+      res.push(`templates.${i}.${value}`);
+    }
+  }
+  return res;
+}
+
+function validateTemplate(template: templateArgs): string[] {
+  const res: string[] = [];
+  if (!template.deploymentName) {
+    res.push("deploymentName is invalid");
```
Comment (jayzhang) [inline packages/fx-core/src/component/driver/arm/validator.ts:65]: Better to extract as localized string

File: packages/fx-core/src/component/driver/arm/validator.ts [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+  }
+
+  if (getFileExtension(template.parameters) !== "json") {
+    res.push("parameters is invalid");
+  }
+  const templateType = getFileExtension(template.path);
+  if (templateType !== TemplateType.Json && templateType !== TemplateType.Bicep) {
+    res.push("path is invalid");
```
Comment (jayzhang) [inline packages/fx-core/src/component/driver/arm/validator.ts:73]: Better to extract as localized string

File: packages/fx-core/src/component/driver/arm/validator.ts (hunk 7 above)
Comment (xzf0587) [reply packages/fx-core/src/component/driver/arm/validator.ts:65]: Here the key name in yaml should not be localized.

File: packages/fx-core/src/component/driver/arm/validator.ts (hunk 8 above)
Comment (xzf0587) [reply packages/fx-core/src/component/driver/arm/validator.ts:73]: Here the key name in yaml should not be localized.

**Review by jayzhang — APPROVED** [review]

**Pull request merged**